GeomapPanel: Add base types to data layer options

drew08t opened the pull request:
What this PR does / why we need it:
Adds base layer types to data layer type selection, including controls for opacity of these data layers. This adds support for OpenSeaMap (and beyond) at the data layer level.

After
(Add XYZ Tile Layer)
![Jun-01-2022 17-47-02](https://user-images.githubusercontent.com/60050885/171524508-2634eab9-93bf-44fd-b932-c9e89c62756d.gif)
(Add ArcGIS MapServer and adjust its opacity)
![Jun-01-2022 17-47-34](https://user-images.githubusercontent.com/60050885/171524519-13311354-be03-426b-99eb-b678125818ea.gif)

Which issue(s) this PR fixes:

Fixes #47812

drew08t committed:
GeomapPanel: Add base types to data layer options

Removed isBaseMap check from data layer filter, which used to remove
base layer types from data layer options. For layer editor, slider to
control opacity is added.

Closes #47812

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=50053) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=50053) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=50053) it.</sub>

drew08t committed:
GeomapPanel: Reorder data layer options

For data layer selection, present user with data layer types first,
before base map layer types. Refactored to consolidate layer options
into a single exported function in registry with a base map boolean.

drew08t committed:
GeomapPanel: Add descriptions to base map types

drew08t commented:
Reordered options with descriptions:
![image](https://user-images.githubusercontent.com/60050885/171752982-eef51baf-0883-4a81-bfa0-1210559e55e2.png)

ryantxu commented on public/app/plugins/panel/geomap/layers/basemaps/generic.ts:
```diff
@@ -20,7 +20,9 @@ export const defaultXYZConfig: XYZConfig = {
   name: 'XYZ Tile layer',
+  description: 'Add map from custom source',
   isBaseMap: true,
+  showOpacity: true,
```
Looking at this now... I think it would be better to have a `hideOpacity` flag, and only set if for 'markers' layer?  (everywhere else opacity seems useful)

drew08t replied:
Changed.

drew08t committed:
Slider: compare value prop with internal state

drew08t committed:
GeomapPanel: add hideOpacity and apply to markers

ryantxu commented on public/app/plugins/panel/geomap/GeomapPanel.tsx:
```diff
@@ -505,6 +505,9 @@ export class GeomapPanel extends Component<Props, State> {
 
     const handler = await item.create(map, options, config.theme2);
     const layer = handler.init();
+    if (options.opacity != null) {
```
maybe also:
```
&& !layer.hideOpacity
```

drew08t replied:
options.opacity is undefined when hideOpacity is true, so we should be fine here.

ryantxu commented on public/app/plugins/panel/geomap/layers/basemaps/esri.ts:
```diff
@@ -57,6 +57,7 @@ export interface ESRIXYZConfig extends XYZConfig {
 export const esriXYZTiles: MapLayerRegistryItem<ESRIXYZConfig> = {
   id: 'esri-xyz',
   name: 'ArcGIS MapServer',
+  description: 'Add map from Arc geographic information system',
```
```suggestion
  description: 'Add layer from an ESRI ArcGIS MapServer',
```

ryantxu commented on public/app/plugins/panel/geomap/layers/basemaps/carto.ts:
```diff
@@ -25,6 +25,7 @@ export const defaultCartoConfig: CartoConfig = {
 export const carto: MapLayerRegistryItem<CartoConfig> = {
   id: 'carto',
   name: 'CARTO reference map',
+  description: 'Add map from CARTO database',
```
```suggestion
  name: 'CARTO basemap',
  description: 'Add layer CARTO Raster basemaps',
```

ryantxu reviewed (approved):
Looks good -- can you change some of the text descriptions?  then lets merge

torkelo commented:
This is a bit confusing, why would you add the map layers as data layers?

drew08t commented:
> I can see why now, makes the category "Data layer" a bit confusing :)

Hmm, perhaps we can call it something a bit more generic.

drew08t committed:
GeomapPanel: update descriptions for map types

ryantxu commented:
> makes the category "Data layer" a bit confusing :)

![wm6VaP3dyl91AdoUMXH2-wx0RkLZqhsIssb4BVHDEas (1)](https://user-images.githubusercontent.com/705951/173166643-92f50ab4-975f-4097-8937-1756f6c21bd6.gif)

Under the hood there is no difference between "Base layer" and "Data layers" -- everything is a list of layers.  I think we should likely update the structure to show this (but not part of this PR)

drew08t committed:
Slider: revert state management
